flit_core/flit_core/config.py

```diff
@@ -822,34 +824,85 @@ def isabs_ish(path):
     return os.path.isabs(path) or path.startswith(('/', '\\'))
 
 
-def normalize_license_expr(s: str):
-    """Validate & normalise an SPDX license expression
+def normalise_compound_license_expr(s: str) -> str:
+    """Validate and normalise a compund SPDX license expression.
 
-    For now this only handles simple expressions (referring to 1 license)
+    Per the specification, licence expression operators (AND, OR and WITH)
+    are matched case-sensitively. The WITH operator is not currently supported.
+
+    Spec: https://spdx.github.io/spdx-spec/v2.2.2/SPDX-license-expressions/
+    """
+    if s == '':
+        raise ConfigError(f"'' is not a recognised SPDX license ID")
+
+    stack = 0
+    parts = []
+    try:
+        for part in filter(None, re.split(r' +|([()])', s)):
+            if part in {'WITH', 'with'}:
+                # provide a sensible error message for the WITH operator
+                raise ConfigError(f"The SPDX 'WITH' operator is not yet supported!")
+            elif part in {'and', 'or', 'with'}:
+                # provide a sensible error message for lowercase operators
+                raise ConfigError(
+                    f"SPDX licence expression operators must be uppercase (saw {s!r})"
+                )
+            elif part in {'AND', 'OR'}:
+                if not parts or parts[-1] in {' AND ', ' OR ', ' WITH ', '('}:
+                    raise ConfigError(f"{s!r} is not a recognised SPDX license ID")
+                parts.append(f' {part} ')
+            elif part == '(':
+                if parts and parts[-1] not in {' AND ', ' OR ', '('}:
+                    raise ConfigError(f"{s!r} is not a recognised SPDX license ID")
+                stack += 1
+                parts.append(part)
+            elif part == ')':
+                if not parts or parts[-1] in {' AND ', ' OR ', ' WITH ', '('}:
+                    raise ConfigError(f"{s!r} is not a recognised SPDX license ID")
+                stack -= 1
+                if stack < 0:
+                    raise ConfigError(f"{s!r} is not a recognised SPDX license ID")
```

Comment:
'License ID' was meant to refer to a single name (the SPDX spec calls these 'license identifiers'). For errors related to the structure of a compound expression, I'd use the term 'license expression', unless we can think of a clearer term for users.

Given that we've already got separate code paths for different error cases, I think it would be useful for the error message to try to describe why. Something like:

```
ConfigError: Invalid SPDX license expression (')' does not follow license ID): '(MIT AND)'
```

Reply:
```
ConfigError: Invalid SPDX license expression: ')' does not follow license ID
    '(MIT AND)'

ConfigError: '(MIT AND)' is not a valid SPDX license expression: ')' does not follow license ID
```

Just trying out some different formats, the extra `()` parentheses in my original suggestion made it a bit hard to scan.

Reply:
@takluyver I've added a reason to each error messages, and in the process fixed two bugs -- one where `'MIT MIT'` was allowed (no operator) and one where `' '` was allowed (whitespace only).